### PR TITLE
Feature/algorand support

### LINKED ISSUE #3: Add support for Algorand

### PR DESCRIPTION
**Pull request by saadibrahim**
closes #3

### CODE REVIEW

File: src/chains/algorand.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            publicKey,
+            address: wallet.addr
+        }
+    }
+
+    static getWallet(mnemonic: string): object {
+        const wallet = algosdk.mnemonicToSecretKey(mnemonic)
+        const privateKey = '0x' + Buffer.from(wallet.sk).toString('hex')
+        const keypair = keyPairFromSecretKey(wallet.sk)
+        const publicKey = '0x' + Buffer.from(keypair.publicKey).toString('hex')
```
Comment (tahpot) [inline src/chains/algorand.ts]: @saadibrahim I'm pretty sure Algorand public keys / addresses aren't represented as `0x` prefixed hex strings.

File: src/chains/algorand.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        return {
+            mnemonic: mnemonic,
+            privateKey,
+            publicKey,
+            address: wallet.addr
+        }
+    }
+
+    static getPublicKey() {
+        throw new Error('Not implemented');
```
Comment (tahpot) [inline src/chains/algorand.ts]: In ETH the address is the public key... is that the same in ALGO?

File: src/chains/algorand.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    static getAddress(privateKey: string): string {
+        const sk = algosdk.secretKeyToMnemonic(privateKeyToArray(privateKey))
+        const wallet = algosdk.mnemonicToSecretKey(sk)
+        return wallet.addr
+    }
+
+    static signTransaction(privateKey: string, transaction: any): object {
```
Comment (tahpot) [inline src/chains/algorand.ts]: Why is this here? It's not part of the `utils.js` interface.

File: src/chains/algorand.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        const {sender, receiver, amount, note, params} = transaction
+        const txn = algosdk.makePaymentTxnWithSuggestedParams(sender, receiver, amount, undefined, note, params);
+        const sk = algosdk.secretKeyToMnemonic(privateKeyToArray(privateKey))
+        const wallet = algosdk.mnemonicToSecretKey(sk)
+        const signedTxn = txn.signTxn(wallet.sk);
+        return {txn, signedTxn}
+    }
+
+    static async signMessage() {
+        throw new Error('Not implemented');
```
Comment (tahpot) [inline src/chains/algorand.ts]: This needs to be implemented. It's a required method on the `utils.js` interface.

File: src/chains/algorand.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    static recoverAddress() {
+        throw new Error('Not implemented');
+    }
+
+    static async verifySignature() {
+        throw new Error('Not implemented');
```
Comment (tahpot) [inline src/chains/algorand.ts]: This needs to be implemented. It's a required method on the `utils.js` interface.

File: src/utils.ts [hunk 6]
```diff
@@ -52,6 +54,10 @@ export default class utils {
         return chains[chain].getAddress(privateKey)
     }
 
+    static signTransaction(chain: string, privateKey: string, transaction: any): object {
```
Comment (tahpot) [inline src/utils.ts]: Why was this added?

**Review by tahpot — CHANGES_REQUESTED** [review]
@saadibrahim I've added review comments for your feedback.

You also need to add unit tests as per the other chains.

File: src/chains/algorand.ts (hunk 2 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: Yes, that's true for ALGO as well.

File: src/chains/algorand.ts (hunk 1 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: yep, I was trynna follow the pattern of the lib, there is no use for public key tho from what I can see. Is it okay to just make this the address?

File: src/chains/algorand.ts (hunk 3 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: Wouldnt we use this lib to sign transactions as well?

File: src/utils.ts (hunk 6 above)
Comment (saadibrahim) [reply src/utils.ts]: since this util does all the offline functions, like creating addresses, signing messages, I figured we would use it to sign transactions too?

File: src/chains/algorand.ts (hunk 4 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: alright. Is this message on slack to do with signing messages:

https://verida-workspace.slack.com/archives/C028EAEQYPJ/p1637637073022400

File: src/chains/algorand.ts (hunk 5 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: related to signMessage, responded above

File: src/chains/algorand.ts (hunk 1 above)
Comment (tahpot) [reply src/chains/algorand.ts]: You'll need to look at the Algorand documentation to see what the public key should be.

The public key is important for other use cases of the library.

Comment (tahpot) [reply src/chains/algorand.ts]: Oh, just saw your next comment. If public key = address, then yes, make them the same.

File: src/chains/algorand.ts (hunk 3 above)
Comment (tahpot) [reply src/chains/algorand.ts]: That wasn't the original intent of the library, but we could.

What's the benefit?

File: src/chains/algorand.ts (hunk 4 above)
Comment (tahpot) [reply src/chains/algorand.ts]: No, that discussion relates to asking other wallets to sign a message -- which they don't support.

In our wallet, we have the private key so we can sign messages properly.

See https://algorand.github.io/js-algorand-sdk/modules.html#signbytes

File: src/chains/algorand.ts (hunk 5 above)
Comment (tahpot) [reply src/chains/algorand.ts]: See `verifyBytes`

File: src/chains/algorand.ts (hunk 1 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: from their documentation:

Screen Shot 2021-12-06 at 4 39 08 PM

It implies the public key is pretty much of no use to the end user, and I do not see how we would be using the public key, the mnenomic, secret/private key and the address is all we need.

File: src/chains/algorand.ts (hunk 3 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: I was looking at the purpose of this library as doing all the "offline" parts, seeing the signMessage method confirmed it for me, but if you see this as something that only creates addresses, that's fair as well and would simplify things but then why is the signMessage stuff in here?

benefit in theory is that we then don't need to call the chain sdks in our code to sign transaction and then just send the signed transaction to the chain.

Edit: for more context, normally this would only be signing the transaction (transaction.signTxn(privateKey)) which would be passed from the app, but way the algosdk works, I am having to do the createTransaction bit here as well, for ethereum we would just be passing the transaction and just handling the signing here.

File: src/chains/algorand.ts (hunk 4 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: Thanks, noted, let me figure this out.

File: src/chains/algorand.ts (hunk 5 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: Thanks

File: src/chains/algorand.ts (hunk 3 above)
Comment (tahpot) [reply src/chains/algorand.ts]: Sure. Let's run with that approach for now and see how it goes. Pretty easy to refactor and remove it later if needed.

File: src/chains/algorand.ts (hunk 1 above)
Comment (tahpot) [reply src/chains/algorand.ts]: If it's easy to get the public key using the AlgoSDK, I would still like to return it for completeness.

Comment (saadibrahim) [reply src/chains/algorand.ts]: It isnt, they dont provide a method. I dove into the sdk code and managed to generate it using nacl but I realized it was unnecessary code for no reason, there was no use for that public key.

**Comment by saadibrahim** [comment]
added `signMessage` and `verifySignature` methods, sorted out the issue I qwas having converting the singature to string and back to Uint8Array, writing unit tests now.

**Comment by saadibrahim** [comment]
@tahpot Added tests, while writing tests I realized signTransaction also relies on another function called getTransactionParams which requires you to connect to the node as it gets something called genesis hash and suggested transaction fee, at this point it just didnt make sense to keep it in here, better to be on the client side.

File: src/chains/algorand.ts (hunk 4 above)
Comment (saadibrahim) [reply src/chains/algorand.ts]: Updated

**Comment by saadibrahim** [comment]
@tahpot You can review it again, have made all the changes requested.

**Comment by nick-verida** [comment]
@tahpot I think this needs a review from you (assuming it is still relevant @saadibrahim?)

**Comment by saadibrahim** [comment]
yes, still very relevant.

**Comment by saadibrahim** [comment]
@tahpot reassigned after adding multi wallet support, its simpler than what you suggested, I can refactor again when working on etheeum.

File: tsconfig.json [hunk 7]
```diff
@@ -6,7 +6,8 @@
       "outDir": "./dist",
       "strict": true,
       "lib": [
-        "es5"
+        "es5",
+        "dom"
```
Comment (tahpot) [inline tsconfig.json]: Why is DOM here? This is a generic library, so shouldn't require the web-browser DOM.

File: package.json [hunk 8]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "@verida/wallet-utils",
-  "version": "1.4.0",
+  "version": "1.5.0",
   "description": "Wallet utilities for managing accounts, keys and signatures across multiple blockchains",
-  "main": "dist/utils.js",
+  "main": "dist/MultiChainWallets.js",
```
Comment (tahpot) [inline package.json]: Why change the entry point for the library?

Can we either have an `index` entry point that exposes both `MultiChainWallets` and `utils` or leave the entry point as `utils` and expose `MultiChainWallets` via `utils`?

We can't lose the ability to create or use a utils for a specific chain.

**Review by tahpot — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 7 above)
Comment (saadibrahim) [reply tsconfig.json]: Screen Shot 2022-01-31 at 12 31 09 PM

^ was getting these typescript errors without it, stackoverflow solution suggested adding dom to fix it but good call out, I just ts-ingored those lined

File: package.json (hunk 8 above)
Comment (saadibrahim) [reply package.json]: Yep, you are right, I have fixed the entry point to be index.js and exposing both utils and multichainwallet

**Comment by tahpot** [comment]
Thanks @saadibrahim, all looks good, feel free to merge.

**Comment by saadibrahim** [comment]
Thanks @tahpot
